File: src/repository/server_client.py
import httpx
from fastapi import HTTPException

SERVER_URL = "http://localhost:8002"

async def server_get(path: str):
    """
    path must begin with /user or /user/ etc.
    Returns JSON body or raises HTTPException for error conditions.
    """
    url = f"{SERVER_URL}{path}"
    try:
        async with httpx.AsyncClient(timeout=5.0) as client:
            resp = await client.get(url)
    except httpx.RequestError as e:
        raise HTTPException(status_code=503, detail=f"Server unreachable: {e}")

    if resp.status_code == 404:
        raise HTTPException(status_code=404, detail="User not found")
    if resp.status_code != 200:
        raise HTTPException(status_code=resp.status_code, detail="Upstream error")

    return resp.json()

File: src/main/java/com/example/demogateway/repository/ServerClient.java
package com.example.demogateway.repository;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Service;
import org.springframework.web.client.ResourceAccessException;
import org.springframework.web.client.RestClientResponseException;
import org.springframework.web.client.RestTemplate;
import org.springframework.web.server.ResponseStatusException;

import java.util.List;
import java.util.Map;

@Service
public class ServerClient {

    @Autowired
    private RestTemplate restTemplate;

    @Value("${upstream.server.url:http://localhost:8002}")
    private String serverUrl;

    public Object serverGet(String path) {
        String url = serverUrl + path;
        try {
            ResponseEntity<Object> resp = restTemplate.getForEntity(url, Object.class);
            if (resp.getStatusCode() == HttpStatus.NOT_FOUND) {
                throw new ResponseStatusException(HttpStatus.NOT_FOUND, "User not found");
            }
            if (resp.getStatusCode() != HttpStatus.OK) {
                throw new ResponseStatusException(resp.getStatusCode(), "Upstream error");
            }
            return resp.getBody();
        } catch (ResourceAccessException e) {
            throw new ResponseStatusException(HttpStatus.SERVICE_UNAVAILABLE, "Server unreachable: " + e.getMessage());
        } catch (RestClientResponseException e) {
            throw new ResponseStatusException(HttpStatus.valueOf(e.getRawStatusCode()), "Upstream error: " + e.getResponseBodyAsString());
        }
    }
}
